src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,175 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { useEffect, useRef, useState } from 'react';
+
+type Props = {
+  todo: Todo;
+  onDelete?: (todoId: number) => void;
+  onloadingTodoIds?: number[];
+  onEdit?: (currentTodo: Todo) => void;
+  isEditingTodo?: boolean;
+  setIsEditingTodo?: (isEditing: boolean) => void;
+  onChecked?: (currentTodo: Todo) => void;
+  errorFromServer?: boolean;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  onloadingTodoIds = [],
+  onDelete = () => {},
+  onEdit = () => {},
+  isEditingTodo,
+  setIsEditingTodo = () => {},
+  onChecked = () => {},
+  errorFromServer,
+}) => {
+  const [title, setTitle] = useState(todo.title);
+  const [editedTodo, setEditedTodo] = useState<Todo | null>(null);
+  const titleField = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (editedTodo && titleField.current) {
+      return titleField.current.focus();
+    }
+  }, [editedTodo]);
+
+  const editTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+  };
+
+  const handleSubmitChanges = async (
+    event: React.KeyboardEvent<HTMLInputElement>,
+  ) => {
+    try {
+      if (event.key === 'Enter') {
+        event.preventDefault();
+        if (!title.trim()) {
+          await onDelete(todo.id);
+
+          return;
+        } else if (title.trim() === todo.title) {
+          setIsEditingTodo(false);
+
+          return;
+        } else {
+          await onEdit({ ...todo, title });
+        }
+      }
+
+      if (event.key === 'Escape') {
+        if (editedTodo) {
+          setTitle(todo.title);
+        }
+      }
+    } catch (error) {
+      setEditedTodo(todo);
+      setIsEditingTodo(true);
+      throw error;
```

Comment:
do not throw an error here - you handled it already, so no need to throw it again. It will lead to a page crash (because you don't try to catch the error thrown here somewhere else)

Reply:
tried to remove it, but the tests fail without this line